_source/_posts/2024-02-27-spring-security-pkce.md

```diff
@@ -0,0 +1,458 @@
+
+Rename `application.properties` to `application.yml` and add the following values:
+
+```yml
+# src/main/resources/application.yml
+server:
+  port: ${PORT}
+
+okta:
+  oauth2:
+    issuer: ${OKTA_OAUTH2_ISSUER}
+    client-id: ${OKTA_OAUTH2_CLIENT_ID}
+    client-secret: ${OKTA_OAUTH2_CLIENT_SECRET}
+
+logging:
+  level:
+    org.springframework.security: DEBUG
+    org.springframework.web: DEBUG
+```
+
+The Okta Spring Boot Starter will detect the presence of the properties above and auto-configure the Spring Security filter chain for OpenID Connect authentication. The configuration also enables security and web logs for analyzing the authentication flow.
+
+Create a `.env` file under the root project directory:
+
+```shell
+touch .env
+```
+
+Populate the `.env` with the following environment variables and values from the previous `auth0 apps create` command output:
+
+```shell
+PORT=4040
+OKTA_OAUTH2_ISSUER=https://<your-auth0-domain>/
+OKTA_OAUTH2_CLIENT_ID=<client-id>
+OKTA_OAUTH2_CLIENT_SECRET=<client-secret>
+```
+
+> The default client authentication method (how the application identifies itself when calling the token endpoint) in Auth0, when you register a regular web application, is `client_secret_post`, but it supports `client_secret_basic` as well. The default client authentication method in Spring Security for confidential clients is `client_secret_basic` if the provider supports it (available in the provider configuration metadata).
+
+Run the application with:
+
+```shell
+./gradlew bootRun
+```
+
+In your browser, open a private navigation window and go to [**http://localhost:4040/**](localhost:8080). Upon clicking the "Log In" button, Spring MVC redirects you to the Auth0 Universal Login page. If you check the application logs, you will see Spring Security redirects to your Auth0 '/authorize' endpoint:
+
+```
+Redirecting to https://dev-avup2laz.us.auth0.com/authorize?response_type=code&client_id=3SgEFDZfV2402hKmNhc0eIN10Z7tem1R&scope=profile%20email%20openid&state=qw8rMCn7N6hv7V4Wx8z5-ejsV8Av8Ypx9KBm5jL6tN4%3D&redirect_uri=http://localhost:4040/login/oauth2/code/okta&nonce=3aiFDmu7aOktibpACDUYuUh4HxLpBAMnVw79EcHDapk&code_challenge=andLN4-6Lu2mtHoPp1Pteu2v87oK_RmzmFLgPaHaY0s&code_challenge_method=S256
+```
+
+As you can see, the last two query parameters in the request to `/authorize` endpoint are _code_challenge_ and _code_challenge_method_.
+
+> [Spring Security](https://docs.spring.io/spring-security/reference/servlet/oauth2/client/authorization-grants.html#_obtaining_authorization) will automatically enable PKCE when `client-secret` is omitted or empty, and `client-authentication-method` is none. A client without a secret is assumed to be a public client.
+
+> Since version 2.1.6, the Okta Starter enables PKCE by default for confidential clients. With the Okta Starter default auto-configuration, PKCE is enabled automatically even if the client-secret is set through the Okta Starter configuration properties.
```

Comment:
with PKCE enabled, what happens if you dont include a client secret? theoritically it should work but I havent tried

Reply:
Added comment,  Auth0 gives 401